drivers/gpio/gpio_andes_atcgpio100.c

```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright (c) 2021 Andes Technology Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file GPIO driver for the AndesTech ATCGPIO100 controller
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <device.h>
+#include <soc.h>
+#include <drivers/gpio.h>
+#include <sys/util.h>
+#include <sys/sys_io.h>
+
+#include "gpio_utils.h"
+
+#define DT_DRV_COMPAT andestech_atcgpio100
+
+/* Andes ATCGPIO100 register definition */
+#define REG_IDR  0x00 /* ID and Revision reg.           */
+#define REG_CFG  0x10 /* Hardware configure reg.        */
+#define REG_DIN  0x20 /* Data In reg.                   */
+#define REG_DOUT 0x24 /* Data Out reg.                  */
+#define REG_DIR  0x28 /* Channel direction reg.         */
+#define REG_DCLR 0x2C /* Data out clear reg.            */
+#define REG_DSET 0x30 /* Data out set reg.              */
+#define REG_PUEN 0x40 /* Pull enable reg.               */
+#define REG_PTYP 0x44 /* Pull type reg.                 */
+#define REG_INTE 0x50 /* Interrupt enable reg.          */
+#define REG_IMD0 0x54 /* Interrupt mode 0 ~ 7 reg.      */
+#define REG_IMD1 0x58 /* Interrupt mode 8 ~ 15 reg.     */
+#define REG_IMD2 0x5C /* Interrupt mode 16 ~ 23 reg.    */
+#define REG_IMD3 0x60 /* Interrupt mode 24 ~ 31 reg.    */
+#define REG_ISTA 0x64 /* Interrupt status reg.          */
+#define REG_DEBE 0x70 /* De-bounce enable reg.          */
+#define REG_DEBC 0x74 /* De-Bounce control reg.         */
+
+#define INT_NO_OPERATION        0x0
+#define INT_HIGH_LEVEL          0x2
+#define INT_LOW_LEVEL           0x3
+#define INT_NEGATIVE_EDGE       0x5
+#define INT_POSITIVE_EDGE       0x6
+#define INT_DUAL_EDGE           0x7
+
+#define PULL_CONFIGURED         (1 << 31)
+#define DEBOUNCE_CONFIGURED     (1 << 29)
+#define DF_DEBOUNCED_SETTING    (0x80000003)
+
+#define DEV_CFG(dev) \
+	((const struct gpio_atcgpio100_config * const)(dev)->config)
+
+#define DEV_DATA(dev) \
+	((struct gpio_atcgpio100_data *)(dev)->data)
+
+
+#define GPIO_CFG(dev)  (DEV_CFG(dev)->base + REG_CFG)
+#define GPIO_DIR(dev)  (DEV_CFG(dev)->base + REG_DIR)
+#define GPIO_DIN(dev)  (DEV_CFG(dev)->base + REG_DIN)
+#define GPIO_DOUT(dev) (DEV_CFG(dev)->base + REG_DOUT)
+#define GPIO_DCLR(dev) (DEV_CFG(dev)->base + REG_DCLR)
+#define GPIO_DSET(dev) (DEV_CFG(dev)->base + REG_DSET)
+#define GPIO_PUEN(dev) (DEV_CFG(dev)->base + REG_PUEN)
+#define GPIO_PTYP(dev) (DEV_CFG(dev)->base + REG_PTYP)
+#define GPIO_INTE(dev) (DEV_CFG(dev)->base + REG_INTE)
+#define GPIO_IMD(dev, idx) (DEV_CFG(dev)->base + REG_IMD0 + (idx << 2))
+#define GPIO_ISTA(dev) (DEV_CFG(dev)->base + REG_ISTA)
+#define GPIO_DEBE(dev) (DEV_CFG(dev)->base + REG_DEBE)
+#define GPIO_DEBC(dev) (DEV_CFG(dev)->base + REG_DEBC)
+
+#define INWORD(x)     sys_read32(x)
+#define OUTWORD(x, d) sys_write32(d, x)
+
+#define SET_GPIO_INT_MODE(cur_val, mode, ch_idx)		\
+		do {						\
+			cur_val &= ~(0x7 << (ch_idx << 2));	\
+			cur_val |= (mode << (ch_idx << 2));	\
+		} while (false)
```

Comment:
I know that `kernel.h` includes `stdbool.h`, but I've seen other files including it separately when using `false`/`true` values, so to keep consistency you probably should do it as well.

Reply:
Done.